internal/pkg/develop/plugin/plugin.go

```diff
@@ -0,0 +1,146 @@
+package plugin
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"path"
+	"text/template"
+
+	pluginTpl "github.com/merico-dev/stream/internal/pkg/develop/plugin/template"
+	"github.com/merico-dev/stream/pkg/util/log"
+)
+
+type Plugin struct {
+	Name string
+}
+
+func NewPlugin(name string) *Plugin {
+	return &Plugin{
+		Name: name,
+	}
+}
+
+// RenderTplFiles takes specified data that the templates needed,
+// then render TplFiles to "Files" and return it as []File.
+func (p *Plugin) RenderTplFiles() ([]pluginTpl.File, error) {
+	retFiles := make([]pluginTpl.File, 0)
+	count := len(pluginTpl.TplFiles)
+	log.Debugf("Template files count: %d.", count)
+
+	for i, tplFile := range pluginTpl.TplFiles {
+		log.Debugf("Render process: %d/%d.", i+1, count)
+		file, err := p.renderTplFile(&tplFile)
+		if err != nil {
+			return nil, err
+		}
+		log.Debugf("File: %v.", *file)
+		retFiles = append(retFiles, *file)
+	}
+
+	return retFiles, nil
+}
+
+// RenderTplFile takes specified data that the template needed,
+// then render one TplFile to File and return it.
+func (p *Plugin) renderTplFile(tplFile *pluginTpl.TplFile) (*pluginTpl.File, error) {
+	name, err := p.renderTplString(tplFile.NameTpl)
+	if err != nil {
+		return nil, err
+	}
+	dir, err := p.renderTplString(tplFile.DirTpl)
+	if err != nil {
+		return nil, err
+	}
+	content, err := p.renderTplString(tplFile.ContentTpl)
+	if err != nil {
+		return nil, err
+	}
+
+	return &pluginTpl.File{
+		Name:    name,
+		Dir:     dir,
+		Content: content,
+	}, nil
+}
+
+// renderTplString get the template string and the data object,
+// then render it and return the rendered string.
+func (p *Plugin) renderTplString(tplStr string) (string, error) {
+	if tplStr == "" {
+		return "", nil
+	}
+
+	t, err := template.New("default").Parse(tplStr)
+	if err != nil {
+		log.Debugf("Template parse failed: %s.", err)
+		log.Debugf("Template content: %s.", tplStr)
+		return "", err
+	}
+
+	var buf bytes.Buffer
+	err = t.Execute(&buf, *p)
+	if err != nil {
+		log.Debugf("Template execute failed: %s.", err)
+		log.Debugf("Template content: %s.", tplStr)
+		log.Debugf("Data object: %v.", *p)
+		return "", err
+	}
+
+	return buf.String(), nil
+}
+
+// PersistFiles gets the []pluginTpl.File, for each File:
+// call the persistFile() method to deal with.
+func (p *Plugin) PersistFiles(files []pluginTpl.File) error {
+	fileCount := len(files)
+	log.Debugf("There are %d files wait to persist.", fileCount)
+	for i, file := range files {
+		log.Debugf("Persist process: %d/%d", i+1, fileCount)
```

Comment:
Missing "period" (.) at the end of the log sentence. Please check for all log lines; but in general it's very good, almost perfect logging. Kudos!

Reply:
Good point. I changed some logs and added some periods before committing. I thought about checking all the logs, but I still forgot. I'll take a check now.